DM-18181: Add butler configuration validation tool

timj opened the pull request:
Adds validation command line tool.

cc/ @MichelleGower

timj committed:
Add registry methods for retrieving all datasetTypes and collection names

timj committed:
Fix exception messages when formatter or template are bad

timj committed:
Add hierarchy of validation methods

Butler can now ask datastores to validate their templates
and formatters.

timj committed:
Add command line tool to validate a butler configuration

timj committed:
Add method to return all dimension values for a dimension

ie ask for all the Instruments

timj committed:
Add more dataset and collection tests

timj committed:
For validation create DatasetRefs for each instrument

This is needed to ensure that we select instrument-specific
overrides.

andy-slac commented on bin.src/validateButlerConfiguration.py:
```diff
@@ -0,0 +1,46 @@
+    # in the configuration then it must be consistent with this collection
+    butler = Butler(config=args.root, collection=args.collection)
+    try:
+        butler.validateConfiguration(logFailures=True)
```
Maybe make `logFailures` configurable via command line arguments? I'm thinking about use case when this can be used just to verify that things are OK (in some other script maybe). `pass` below can be replaced with `sys.exit(1)` to make useful like that.

andy-slac commented on python/lsst/daf/butler/butler.py:
```diff
@@ -39,13 +39,18 @@
 
-
-__all__ = ("Butler",)
+__all__ = ("Butler", "ButlerValidationError")
```
`__all__` should be above all imports.

timj replied:
Ok. (even though I didn't put it there to begin with)

andy-slac commented on python/lsst/daf/butler/butler.py:
```diff
@@ -546,3 +551,42 @@ def remove(self, datasetRefOrType, dataId=None, *, delete=True, remember=True, *
+        entities = list(self.registry.getAllDatasetTypes())
+
+        # Find all the registered instruments
+        instrument = []
```
`instruments`?

andy-slac commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -365,3 +372,27 @@ def transfer(self, inputDatastore, datasetRef):
         raise NotImplementedError("Must be implemented by subclass")
+
+    @abstractmethod
+    def validateConfiguration(self, *entities, logFailures=False):
```
What is the reason for using `*entities`, can't you just say that `entities` is a list or iterable?

timj replied:
I wanted to be able to say `datastore.validateConfiguration(ref)` without having to put the one DatasetRef inside a list.  I can change it if you think it's cleaner that way.

andy-slac replied:
If you want to accept both single item and a set of items in the same method then I think passing it via single argument as an iterable is a cleaner way. Alternatively you can accept both and wrap single items into a list internally (checking argument type) but I find that sort of interface a bit fragile, especially when item type is not fixed. It's your call but I'd prefer iterable here if it does not make API more difficult to use.

andy-slac commented on python/lsst/daf/butler/core/fileTemplates.py:
```diff
@@ -90,29 +91,49 @@ def __init__(self, config, default=None):
+            if len(failed) == 1:
+                msg = str(failed[0])
+            else:
+                msg = f"{len(failed)} template validation failures: {'; '.join([str(f) for f in failed])}"
```
To my taste it's a bit too much for f-string expression, I'd probably move that `join()` into a regular code. And if `failed` contained strings instead of exceptions then you would not need list comprehension here.

andy-slac commented on python/lsst/daf/butler/core/registry.py:
```diff
@@ -349,6 +360,17 @@ def getDatasetType(self, name):
+        Returns
+        -------
+        types : `tuple` of `DatasetType`
+            All the `DatasetType`\ s.
```
I'd drop backticks here (just write DatasetTypes), there is plenty of links already to `DatasetType` in this docstring.

timj replied:
I've tweaked the wording.

andy-slac commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -365,3 +372,27 @@ def transfer(self, inputDatastore, datasetRef):
+
+    @abstractmethod
+    def validateConfiguration(self, *entities, logFailures=False):
+        """Validate some of the configuration for this datastore.
```
What do you mean by "some of the configuration"? Parts of configuration related to entities? Or is it entirely up to a subclass?

timj replied:
It's entirely up to the subclass and I'm only checking that templates and formatters are consistent. Even that is only done by PosixDatastore. InMemoryDatastore doesn't check anything at the moment.  The code isn't looking at every configuration item.

andy-slac commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -456,3 +457,40 @@ def transfer(self, inputDatastore, ref):
         return [datastore.put(inMemoryDataset, ref) for datastore in self.datastores]
+
+    def validateConfiguration(self, *entities, logFailures=False):
+        """Validate some of the configuration for this datastore.
```
I think we starting to use more of

        # Docstring inherited from Datastore.validateConfiguration

style of comments instead of copying whole docstrings from base classes?

timj replied:
Agreed, but here the notes section is different because it explains that each datastore is queried in turn.

andy-slac replied:
I wish we could extend inherited docstrings. Or could we do it with something like

    See `BaseClass.method` for description of parameters and return types.

    Notes
    -------
    Some additional info

?

andy-slac commented on python/lsst/daf/butler/datastores/posixDatastore.py:
```diff
@@ -560,6 +558,53 @@ def transfer(self, inputDatastore, ref):
+            if templateFailed:
+                messages.append(str(templateFailed))
+            if formatterFailed:
+                messages.append(",".join(formatterFailed))
```
Do you need to wrap each item in `formatterFailed` with `str()`?

andy-slac commented on python/lsst/daf/butler/registries/sqlRegistry.py:
```diff
@@ -267,6 +267,14 @@ def _makeDatasetRefFromRow(self, row, datasetType=None, dataId=None):
+    def getAllCollections(self):
+        # Docstring inherited from Registry.getAllCollections
+        datasetCollectionTable = self._schema.tables["DatasetCollection"]
+        result = self._connection.execute(select([datasetCollectionTable.c.collection]).distinct()).fetchall()
```
I started to dislike that single-letter `c` attribute, but it's OK to keep it for consistency with the rest of the code in this file.

andy-slac commented on python/lsst/daf/butler/core/registry.py:
```diff
@@ -349,6 +360,17 @@ def getDatasetType(self, name):
+
+        Returns
+        -------
+        types : `tuple` of `DatasetType`
```
Why is it a tuple? (I'm probably an old school guy where tuple means a "record", and list means a sequence).

timj replied:
I don't really like returning mutable object for something that is meant to reflect a fixed state. I think of tuple as a read-only list.

andy-slac replied:
Could you return `frozenset` instead?

timj replied:
Yes. Not sure why I don't.

andy-slac commented on tests/test_sqlRegistry.py:
```diff
@@ -449,6 +455,20 @@ def testDatasetUnit(self):
+
+        instrumentEntries = registry.findDimensionEntries(dimensionName)
+        instruments = {e["instrument"] for e in instrumentEntries}
+        self.assertEqual(instruments, set(["DummyCam", "DummyCam2"]))
```
You do `{item, item}` in other places but `set([item, item])` here?

andy-slac reviewed (approved):
Looks OK, few minor comments.

timj committed:
Improve LookupKey stringification

If only a name is defined, return that name without embellishment.

timj committed:
Stringify keys that fail rather than the list

If the list of keys is stringified then repr() is used to
stringify the LookupKeys. This is not very human readable
so stringify each key in turn.

timj committed:
Ensure that exception is stringified before including in message

timj committed:
Fallback to dataId keys in template validation

The dataId keys must still be valid parts of the template
even if dimension links are not available.

timj committed:
Add datastore tests for config validation

timj committed:
Return frozenset for getAllDatasetTypes instead of a tuple

timj committed:
Change validation interface to take single iterable rather than *args

timj committed:
Update validation script to use bad exit status and allow quiet mode

timj committed:
Clean up file template validation error message

timj committed:
Consistently use set constructor {}

timj committed:
Be consistent with __all__ placement

Put at top of file

timj committed:
File is part of daf_butler not ci_hsc

timj committed:
Add ability to override dataset type names in validation

timj committed:
Allow for comma-separated values on command line

timj committed:
For validation the dataset overrides can be empty lists

Not solely None

timj committed:
Add method to indicate whether this DatasetType is a component

timj committed:
Check that a component field is present for a component DatasetType

timj committed:
Add component to the default template

Needed now that datasetType only includes the composite name.

timj committed:
If ignoring a composite dataset, ignore all components

timj committed:
Add some newlines in exception message to make it easier to read

timj committed:
Add tests for ignoring and specific dataset type validation

timj committed:
Refactor template/formatter retrieval to enable access to match key

This allows for more useful validation error reports.

timj committed:
Clean up LookupKey internals to make for cleaner stringification

timj committed:
Fix DatasetType.isComponent and add tests for it

timj committed:
Use ValidationError if a Dimension is not part of a Universe

Previously a KeyError was used.

timj committed:
Add validation of all keys in templates and formatter configurations

No check every key even if there is no registered dataset type /
dimension / storage class  combination that would select that key.

This allows for typos to be spotted.

timj committed:
Fix datasetType override logic when using the validation script

timj committed:
Complain if component is mandatory in template but not a composite

timj committed:
Forward the error message up one level during validation

andy-slac commented on python/lsst/daf/butler/butler.py:
```diff
@@ -546,3 +551,141 @@ def remove(self, datasetRefOrType, dataId=None, *, delete=True, remember=True, *
+        if datastoreErrorStr:
+            messages.append(datastoreErrorStr)
+
+        for failed, msg in ((failedNames, "Keys with out corresponding DatasetType or StorageClass entry: "),
```
with out or without?

andy-slac commented on python/lsst/daf/butler/core/configSupport.py:
```diff
@@ -75,15 +75,23 @@ def __init__(self, name=None, dimensions=None, dataId=None):
-        # tuples.
+        # tuples so that it is not mutable
         if dataId is not None:
             self._dataId = frozenset((k, v) for k, v in dataId.items())
```
Is comprehension needed here, I'd think `frozenset(dataId.items())` should do the same?

andy-slac commented on python/lsst/daf/butler/core/mappingFactory.py:
```diff
@@ -110,8 +140,35 @@ def getFromRegistry(self, *targetClasses):
+        Raises
+        ------
+        KeyError
+            None of the supplied target classes match an item in the registry.
```
"Raised if"

andy-slac commented on python/lsst/daf/butler/butler.py:
```diff
@@ -546,3 +551,141 @@ def remove(self, datasetRefOrType, dataId=None, *, delete=True, remember=True, *
+                        datasetType = self.registry.getDatasetType(key.name)
+                    except KeyError:
+                        if logFailures:
+                            log.fatal(f"Key '{key}' does not correspond to a DatasetType or StorageClass")
```
Does not flake8 warn that you are supposed to use `%`-style formatting with logging? Or is it confused by f-strings?

andy-slac commented on python/lsst/daf/butler/core/fileTemplates.py:
```diff
@@ -80,42 +84,93 @@ def __init__(self, config, default=None):
+            except KeyError as e:
+                failed.append(e)
+                if logFailures:
+                    log.fatal("%s", str(e))
```
`str` is superfluous here (and few other `log.fatal` cases too)

andy-slac reviewed (approved):
Looks OK (but I'm under jetlag+coffee so things look mostly blurry 🙂). Few trivial comments below.

timj committed:
Remove old key before adding new one

The new and old hash to the same value because LookupKey
equality only looks at dimensions and not DimensionGraph
vs DimensionsNameSet.

timj committed:
Fix error message reporting for multiple failed validation templates

timj committed:
In second validation pass only check key validity

This prevents double checking of values. In most cases
the checking of values is not relevant if the key is not
used.

timj committed:
Issue log message for keys that never got checked during validation

timj committed:
Use MappingProxy in FileTemplates so that dict is read only

timj committed:
Add some more template validation tests

timj committed:
Fix error message stringification

KeyError turns out to be a special case.

timj committed:
Construct frozenset directly from items without a comprehension

timj committed:
Update Raises documentation to use Raised if...

timj committed:
Fix failure message text

timj committed:
Do not use f-strings in log messages

timj committed:
Fix long docstring lines
